### DISCUSSION TITLE
Puffin-D accepting only 99999 bases as input

### DISCUSSION THREAD

**Discussion by BrettAdey** [jzhoulab/puffin#18] [Q&A] [answered]
Hi all,

When I run Puffin-D using a fasta input of exactly 100,000 bases, I receive an error saying the input is too large (100,001 bases).

However, when I run it with 99,999, Puffin-D works as expected and produces a 1x10x100kb output.

I created the input with the following code, chunking the human genome:

Chunk T2T genome with step size & window size of 99,999. Then remove all chunks containing "N"s and use width 0 to put the whole sequence on a single line
`seqkit sliding -s 99999 -W 99999 T2T.fasta | seqkit grep -s -r -p "N" -v |  seqkit seq -w 0 > T2T_100kbchunks.fasta`

After splitting this by chromosome, I confirmed my inputs are 99,999b with:

`awk 'FNR==2 {print FILENAME, length($0)}' *.fasta`

NC_060925.1.fasta 99999
NC_060926.1.fasta 99999
NC_060927.1.fasta 99999
NC_060928.1.fasta 99999
NC_060929.1.fasta 99999
NC_060930.1.fasta 99999
NC_060931.1.fasta 99999
...

and also with:
`
grep '^>' NC_060925.1.fasta | wc -l`
2483 headers (so that many ~100kbs in T2T Chr 1)

`grep -v '^>' NC_060925.1.fasta | tr -d '\n' | wc -c`
248297517 (Total number of non-header characters (i.e. nucleotides))

248297517 ÷ 2483 = **99999 (nucleotides per contig)**

I confirmed the Puffin output is indeed 100,000 in Python with puffin_output.shape.

Could anyone please explain how Puffin-D seems to only accept as input a sequence that is one base shorter than it predicts? I'm sure there's something obvious I've missed (perhaps relating to 0-based indexing, hidden new-line characters, or inclusive genome start-stop coordinates), but I can't find what it could be.

Many thanks,
Brett

**Comment by KseniiaDundyk** [comment, marked as answer]
Hi Brett,
This issue was fixed. It occurred because the code treated /n at the end of the sequence as a symbol, and it was converted to the [0.25, 0.25, 0.25, 0.25] encoding.

Best regards,
Kseniia Dudnyk

**Comment by BrettAdey** [comment]
Hi Kseniia,

Thank you for getting back to me - I've updated the Python script and it's
working as expected for me now. Much appreciated!

All the best,
Brett

On Thu, 14 Nov 2024 at 10:12, Kseniia Dudnyk ***@***.***>
wrote:

> Hi Brett,
> This issue was fixed. It occurred because the code treated /n at the end
> of the sequence as a symbol, and it was converted to the [0.25, 0.25, 0.25,
> 0.25] encoding.
>
> Best regards,
> Kseniia Dudnyk
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jzhoulab/puffin/discussions/18#discussioncomment-11246397>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALP32PJTHHKUXGO4P7WCHSD2AO6EBAVCNFSM6AAAAABQ7RBJA6VHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTCMRUGYZTSNY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
*Brett Adey*

*PhD Student*
*School of Biological Sciences | Te Kura Mātauranga Koiora*
*University of Auckland **| Te Whare Wānanga o Tāmaki Makaurau*
